### ISSUE TITLE
Misplaced removeNulls in XSS rules

### ISSUE THREAD

**Issue by eff7** [SpiderLabs/owasp-modsecurity-crs#582]
Hi,

I am referring to OWASP ModSecurity Core Rule Set ver.3.0.0 (currently at rc2)

The xss prevention rules in rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf contain t:removeNulls as an early transformation action.
It seems the location of this transformation is incorrect.

Rule sample:

```
#
# -=[ XSS Filters - Category 1 ]=-
# http://xssplayground.net23.net/xssfilter.html
# script tag based XSS vectors, e.g., <script> alert(1)</script>
#
SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|REQUEST_HEADERS:User-Agent|REQUEST_HEADERS:Referer|ARGS_NAMES|ARGS|XML:/* "(?i)(<script[^>]*>[\s\S]*?)" \
	"msg:'XSS Filter - Category 1: Script Tag Vector',\
	id:941110,\
	phase:request,\
	severity:'CRITICAL',\
	rev:'2',\
	ver:'OWASP_CRS/3.0.0',\
	maturity:'4',\
	accuracy:'9',\
	t:none,t:removeNulls,t:utf8toUnicode,t:urlDecodeUni,t:htmlEntityDecode,t:jsDecode,t:cssDecode,\
	block,\
	ctl:auditLogParts=+E,\
	capture,\
	tag:'application-multi',\
	tag:'language-multi',\
	tag:'platform-multi',\
	tag:'attack-xss',\
	tag:'OWASP_CRS/WEB_ATTACK/XSS',\
	tag:'WASCTC/WASC-8',\
	tag:'WASCTC/WASC-22',\
	tag:'OWASP_TOP_10/A3',\
	tag:'OWASP_AppSensor/IE1',\
	tag:'CAPEC-242',\
	logdata:'Matched Data: %{TX.0} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
	setvar:'tx.msg=%{rule.msg}',\
	setvar:tx.xss_score=+%{tx.critical_anomaly_score},\
	setvar:tx.anomaly_score=+%{tx.critical_anomaly_score},\
	setvar:tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/XSS-%{matched_var_name}=%{tx.0}"
```

The attack vector will use url parameters that contain an embedded encoded null character that is later ignored by the browser. E.g.: `<scr%00ipt>`
So remove nulls should be performed _after_ the argument is decoded to allow the rule to catch the offending pattern.

In our tests we put t:removeNulls as the last transformation and it gives the desired results.

```
	t:none,t:utf8toUnicode,t:urlDecodeUni,t:htmlEntityDecode,t:jsDecode,t:cssDecode,t:removeNulls,\
```

**Comment by dune73** [comment]
Moving the transformation makes a lot of sense, in my eyes. Feel like writing a pull request?

What do the others think?

**Comment by csanders-git** [comment]
Moving the transformation seems to be the only option.

**Comment by csanders-git** [comment]
Closed in favor of PR #584
